### PR TITLE
Issue #2861181 : Better handling of customer remote IDs for payments.

### PR DESCRIPTION
**Pull request by vasike**

### CODE REVIEW

File: modules/payment/commerce_payment.post_update.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * Post update functions for Payment.
+ */
+
+/**
+ * Update user Remote ID field data using payments gateways instead of module.
+ * @see https://www.drupal.org/node/2861181
+ */
+function commerce_payment_post_update_1(&$sandbox) {
+  // Get the user ids that have 'commerce_remote_id' field not emtpy.
+  $result = \Drupal::entityQuery('user')->exists('commerce_remote_id')->execute();
```
Comment (mglaman) [inline modules/payment/commerce_payment.post_update.php]: Wouldn't we only want to run this the first time, and I think we could just do a count query

File: modules/payment/commerce_payment.post_update.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    $payment_gateway_data = $payment_gateway->toArray();
+    $module = reset($payment_gateway_data['dependencies']['module']);
+    if (!isset($payment_gateways_by_module[$module])) {
+      $payment_gateways_by_module[$module] = $payment_gateway;
+    }
+  }
+
+  // Process entities by groups of 20.
+  $limit = 20;
+  $result = array_slice($result, $sandbox['current'], $limit);
```
Comment (mglaman) [inline modules/payment/commerce_payment.post_update.php]: Why not put a range on the query instead

**Review by mglaman — COMMENTED** [review]
Just some concerns around the post update function.

File: modules/payment/commerce_payment.post_update.php (hunk 1 above)
Comment (vasike) [reply modules/payment/commerce_payment.post_update.php]: is it necessary?

File: modules/payment/commerce_payment.post_update.php (hunk 2 above)
Comment (vasike) [reply modules/payment/commerce_payment.post_update.php]: is it necessary?

File: modules/payment/commerce_payment.post_update.php (hunk 1 above)
Comment (bojanz) [reply modules/payment/commerce_payment.post_update.php]: A count query would probably be more efficient, let's do that.

Comment (vasike) [reply modules/payment/commerce_payment.post_update.php]: it should be done now

File: modules/payment/commerce_payment.post_update.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * @see https://www.drupal.org/node/2861181
+ */
+function commerce_payment_post_update_1(&$sandbox) {
+  // Get the user ids that have 'commerce_remote_id' field not empty.
+  // First check is any user has this field set.
+  if (\Drupal::entityQuery('user')->exists('commerce_remote_id')->count()->execute() == 0) {
+    return t('No user Remote ID field data to update.');
+  }
+  $result = \Drupal::entityQuery('user')->exists('commerce_remote_id')->execute();
```
Comment (bojanz) [inline modules/payment/commerce_payment.post_update.php]: We should run the count query only under if (!isset($sandbox['current'])) {
After that it's cached in $sandbox.

Then run the real query with a range each time.

File: modules/payment/commerce_payment.post_update.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    $sandbox['max'] = count($result);
+    // A place to store messages during the run.
+  }
+
+  // Get the first payment gateway for every payment module.
+  $payment_gateways_by_module = [];
+  /** @var \Drupal\commerce_payment\PaymentGatewayStorageInterface $payment_gateway_storage */
+  $payment_gateway_storage = \Drupal::entityTypeManager()->getStorage('commerce_payment_gateway');
+  $payment_gateways = $payment_gateway_storage->loadMultiple();
+  uasort($payment_gateways, [$payment_gateway_storage->getEntityType()->getClass(), 'sort']);
```
Comment (bojanz) [inline modules/payment/commerce_payment.post_update.php]: You should be able to use $payment_gateway_storage->getEntityType()->getProvider() to get the module.

File: modules/payment/src/Entity/PaymentMethod.php [hunk 5]
```diff
@@ -224,6 +224,29 @@ public function setCreatedTime($timestamp) {
   /**
    * {@inheritdoc}
    */
+  public function getOwnerRemoteId() {
```
Comment (bojanz) [inline modules/payment/src/Entity/PaymentMethod.php]: Is there a point in having these methods? We usually need the remote customer ID before creating a payment method, so we need them on PaymentGatewayBase instead.

**Review by bojanz — CHANGES_REQUESTED** [review]

File: modules/payment/src/Entity/PaymentMethod.php (hunk 5 above)
Comment (vasike) [reply modules/payment/src/Entity/PaymentMethod.php]: i also thought the same initially. but working on it looks like PaymentMethod entity methods

File: modules/payment/commerce_payment.post_update.php (hunk 3 above)
Comment (vasike) [reply modules/payment/commerce_payment.post_update.php]: should be done

File: modules/payment/commerce_payment.post_update.php (hunk 4 above)
Comment (vasike) [reply modules/payment/commerce_payment.post_update.php]: no idea, what you mean

**Comment by bojanz** [comment]
Committed in 14786dd4afa8e94e5246199759e10ff19ac0f348.

**Pull request closed without merging**